### PR TITLE
Retry docker event subscription

### LINKED ISSUE #2016: Weave proxy losing docker event updates
As mentioned at https://github.com/weaveworks/weave/issues/1775#issuecomment-190734573, [these logs](https://gist.github.com/errordeveloper/8058d847572e1786bce0) show a proxy failing to receive Docker events, and subsequent investigation shows the goroutine for `fsouza/go-dockerclient` missing, presumed exited because of unexpected error.

### PR DESCRIPTION
**Pull request by awh**
Fixes #2016.

### CODE REVIEW

File: common/docker/client.go [hunk 1]
```diff
@@ -72,21 +78,29 @@ func (c *Client) Info() string {
...
 	go func() {
-		for event := range events {
-			switch event.Status {
-			case "start":
-				id := event.ID
-				ob.ContainerStarted(id)
-			case "die":
-				id := event.ID
-				ob.ContainerDied(id)
+		retryInterval := InitialInterval
```
Comment (bboreham) [inline common/docker/client.go]: wouldn't we want to reset the retry interval on successful connection?

Comment (awh) [inline common/docker/client.go]: > wouldn't we want to reset the retry interval on successful connection?

As it stands, `AddEventListener` can only return an error if you try to subscribe the same listener twice - the actual docker socket dial and `GET /events` request happen asynchronously, and the only way we learn about either of them failing is via event channel closure.

Consequently, in this case, 'successful connection' equates to the `for ... range events` _blocking_, which does not give us a handy place to reset the interval. I did give it some thought, the best I could come up with was to record the amount of time we spend in the event loop and if it exceeds the retryInterval when the channel closes we reset. Thoughts?

**Comment by awh** [comment]
> record the amount of time we spend in the event loop and if it exceeds the retryInterval when the channel closes we reset

@bboreham implemented - PTAL

File: common/docker/client.go (hunk 1 above)
Comment (bboreham) [inline common/docker/client.go]: It's not bad. Consider if we fail to connect a number of times, and the retry interval goes up to  1 minute.  Then we connect, but it fails again after 30 seconds.  We will now wait 1 minute before trying again.

However, having had various long pauses while we retried, this doesn't seem much more terrible.

I guess the back-off is just to stop killing the machine with rapid reconnects.

File: common/docker/client.go [hunk 2]
```diff
@@ -4,12 +4,18 @@ import (
...
+	"time"
 
 	docker "github.com/fsouza/go-dockerclient"
 
 	. "github.com/weaveworks/weave/common"
 )
 
+const (
+	InitialInterval = 1 * time.Second
+	MaxInterval     = 60 * time.Second
```
Comment (rade) [inline common/docker/client.go]: 60 seems high. 10? 20?

Comment (bboreham) [inline common/docker/client.go]: message posted simultaneously with merge

**Pull request merged**